clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -3141,6 +3148,10 @@ def warn_attribute_ignored_no_calls_in_stmt: Warning<
   "statement">,
   InGroup<IgnoredAttributes>;
 
+def warn_attribute_needs_aggregate : Warning<
```

Comment:
This seems like it should be an error here, mis-use of this should be an error IMO.

Reply:
I still disagree, in that misuse of this attribute should be harsher than a warning, we have precedent for it.  As far as the 'older compiler' thing, typically that is the point of version checks/etc.

That said, I'm LESS Sure than I was before, so I wouldn't mind seeing Aaron's opinion here.

Reply:
just a quick grepping (and I gave up pretty quickly :D :

`err_anyx86_interrupt_attribute` < Though we're super inconsistent with this attribute
`err_using_if_exists_on_ctor` < This is almost exactly the current situation
`err_objc_direct_on_protocol`
`err_destroy_attr_on_non_static_var`
`err_musttail_needs_prototype`
`err_musttail_forbidden_from_this_context`
`err_musttail_needs_call`
`err_attribute_no_member_function` < Directly related, appertainment
`err_alignas_attribute_wrong_decl_type` < also appertainment

Counter points:
`warn_pure_function_returns_void`
`warn_objc_direct_property_ignored`
`warn_attribute_ignored_no_calls_in_stmt`
`warn_attribute_ignored_non_function_pointer`
`warn_function_attribute_ignored_in_stmt`

So it seems that we're pretty inconsistent, but 'appertainment on entity type' SEEMS to lean towards error?  So my examples have managed to leave me 'whelmed' with evidence for each way.  I still think waiting on Aaron is a good bit here, he was Attribute code owner for a long time, so he probably has more context for comment.

Reply:
Another place for Aaron, whether `warn_field_requires_explicit_init` and `warn_attribute_needs_aggregate` should be errors, or can remain warnings.

Reply:
Thank you for your patience, sorry it took so long to get to this PR in my queue!

I think I come down on the side of `warn_attribute_needs_aggregate` being a warning rather than an error. We have a direct recovery path (ignore the attribute, proceed as normal, which is what every other compiler will do anyway), but also, this leaves the door a bit more open to supporting non-aggregates in the future because it means users can put the attribute on a non-aggregate in Clang 21 and it doesn't fail compilation in Clang 20, without requiring conditional compilation.

`warn_field_requires_explicit_init` seems more defensible as an error because that's "you're misusing this". But if it's a hard error, I don't know what kind of escape hatch we would give users who are getting a type declared in a header they can't modify but they somehow have a legit use case without naming the field. (Maybe they're doing partial initialization and filling out the rest in subsequent statements, etc) So maybe a warning that defaults to an error is a better approach because you still get the strong encouragement to not do that, but you have an escape hatch (pragmas if you want to only silence once instance, command line flag otherwise). WDYT?

Reply:
Okay, that seems reasonable to me, thanks!